2.metrics_ablation_analysis/src/image_ablation_analysis/ablation_runner.py

```diff
@@ -0,0 +1,255 @@
+        `run()` method to allow for applying multiple ablations per each image.
+    - Writes sidecar JSON files with augmentation metadata with the same file
+        name as the ablated image to facilitate later analysis
+        (in case index breaks).
+    """
+
+    def __init__(
+        self,
+        images_root: Union[str, Path],
+        ablation_root: Union[str, Path],
+        loaddata_csvs: List[Path | str],
+        *,
+        index_filename: str = "ablated_index",
+        keep_meta_columns: Optional[List[str]] = None,
+        suffixes: Optional[Iterable[str]] = TIFF_SUFFIXES,
+        writer: Optional[Callable[[Path, Any], None]] = None,
+        skip_if_indexed: bool = True,
+        dry_run: bool = False,
+    ):
+        """
+        Initialize AblationRunner.
+
+        :param images_root: Root folder containing source images.
+        :param ablation_root: Root folder to write ablated images.
+        :param loaddata_csvs: List of CSV files defining source images and metadata.
+        :param index_filename: Filename for the Parquet index under ablation_root.
+        :param keep_meta_columns: List of metadata columns to keep from loaddata.
+            If None, keeps all columns present in loaddata.
+        :param suffixes: Iterable of file suffixes (including dot) to consider as images.
+        :param writer: Optional callable to write images. If None, uses tifffile.imwrite.
+        :param skip_if_indexed: If True, skip images already present in index.
+        :param dry_run: If True, do not write any files; only simulate.
+        """
+        self.images_root = check_path(images_root, ensure_dir=True)
+        self.ablation_root = check_path(ablation_root, ensure_dir=True)
+        
+        self.suffixes = suffixes
+        self.writer = writer or tiff.imwrite
+
+        self.loaddata = LoadDataIndex(
+            [Path(p) for p in loaddata_csvs]
+        )
+
+        self.index_filename = index_filename
+        self.index_path = self.ablation_root / self.index_filename
+        self.index = ParquetIndex(self.index_path)
+        self._default_meta_cols = self.loaddata.columns
+        self.keep_meta_columns = keep_meta_columns or self.loaddata.columns
+
+        self.skip_if_indexed = skip_if_indexed
+        self.dry_run = dry_run        
+
+    def run(
+        self,
+        *,
+        augment_hook: Callable[[Path], Iterator[AugVariant]],
+        writer: Optional[callable[[Path, Any], None]] = None,
+        run_id: Optional[str] = None,
+    ) -> None:
+        """
+        Run ablation analysis using the provided augment_hook.
+
+        :param augment_hook: Callable that takes a source image path and
+            yields AugVariant instances.
+        :param writer: Optional callable to write images. 
+            If None, uses default tifffile
+        :param run_id: Optional run identifier. If None, generates a UUID. 
+        """
+
+        writer = writer or self._default_tiff_writer
```

Comment:
Why is the `writer` parameter explicitly required here? If a user initializes the class with a custom writer, `self.writer` should already reference that writer. However, if the function is called without specifying `writer`, it instead uses `self._default_tiff_writer`, which ends up overriding or ignoring the user-provided writer. This could be confusing or unintuitive for users who expect their custom writer to be retained.

Would it make more sense to default to `self.writer` instead of requiring the `writer` argument? For example, something like:


```suggestion
        writer = self.writer or self._default_tiff_writer
```

could be that I am misunderstanding the purpose of the `writer` parameter

Reply:
My bad, writer should not be in `__init__` but be specified on a per `__call__` basis to allow for flexibility, plus the default self.writer initialization is kind of shaky (as  a result of me deciding to use per call writer and forgetting to remove older code).